읽기 동작 기능 추가

dkdud9261 opened the pull request:
## 내용
### 1. 읽기
- 읽기 단건 조회

### 2. 페이지 넘기기
- 해당 책 페이지 수 넘으면 예외
- 끝난 읽기면 예외

### 3. 감상 기록
- 한건 조회 : 단건 조회에 필드 포함
- 수정 : 최대 글자 수 넘으면 예외

### 4. 읽기 마침
- 끝난 읽기면 예외

## 참고
- 읽기 달력 추가 시 페이지 넘기기 로직 다시 생각해 봐야함 ...

dkdud9261 committed:
Feat: 읽기 단건 조회 api 추가

dkdud9261 committed:
Fix: 알라딘 api 요청 시 uri 생성 오류 수정

dkdud9261 committed:
Feat: 페이지 넘기기 api 추가

dkdud9261 committed:
Fix: SecurityConfig 오타 수정

dkdud9261 committed:
Feat: 페이지 넘기기 진행 중인 읽기인지 검사 추가

dkdud9261 committed:
Feat: 감상 기록 수정 api 추가

dkdud9261 committed:
Feat: 기록 조회 응답 데이터에 감상문 필드 추가

dkdud9261 committed:
Feat: 읽기 마침 api 추가

dkdud9261 committed:
Fix: records api 호출 에러 해결

- security filter chain에 CSRF disable 추가
- request body form(dto)에 기본 생성자 추가
- @RequestMapping 제거

dkdud9261 committed:
Feat: regdate 저장

dkdud9261 committed:
Fix: recordDTO 생성자 수정

PPakSang commented on ilkda/src/main/java/com/ilkda/server/record/controller/RecordController.java:
```diff
@@ -12,25 +14,57 @@
         );
     }
+
+    @GetMapping("/api/v1/records/{id}")
+    public SuccessResponse<RecordDTO> getRecordReading(@PathVariable Long id) {
+        Record record = recordService.getRecordReading(id);
+        return new SuccessResponse<>(
+                RecordDTO.builder()
+                        .id(record.getId())
+                        .title(record.getBook().getTitle())
+                        .author(record.getBook().getAuthor())
+                        .bookPage(record.getBook().getPage())
+                        .readPage(record.getReadPage())
+                        .text(record.getText())
+                        .build()
+        );
+    }
```
Record 객체를 바탕으로 RecordDTO 객체를 생성하는 팩토리 메소드를 RecordDTO 내부에 추가하면 좋을 것 같습니다.

dkdud9261 replied:
RecordDTO에 static 메소드로 추가했습니다~!

PPakSang commented on ilkda/src/main/java/com/ilkda/server/record/controller/RecordController.java:
```diff
@@ -12,25 +14,57 @@
     public SuccessResponse<Long> createRecord(@RequestBody RegisterRecordForm form,
                                               Authentication authentication) {
-        Long memberId = (Long) authentication.getPrincipal();
+        Long memberId = Long.parseLong((String) authentication.getPrincipal());
```
Authentication 에서 memberId 를  추출하는 로직은 인증권한이 필요한 영역에서 모두 사용될 것 같습니다. 따라서 별도의 클래스 정적 메소드로의 분리도 고려해보면 좋을 것 같습니다.

dkdud9261 replied:
@AuthenticationPrincipal 어노테이션으로 UserDetails를 사용하는 방식으로 수정했습니다~!

PPakSang commented on ilkda/src/main/java/com/ilkda/server/record/controller/RecordController.java:
```diff
@@ -12,25 +14,57 @@
+        );
+    }
+
+    @PutMapping("/api/v1/records/{id}/page")
+    public SuccessResponse<Long> updateReadPage(@PathVariable Long id,
+                                                @RequestParam Long page) {
+        return new SuccessResponse<>(recordService.updateReadPage(id, page));
+    }
```
보편적으로 update 를 위한 데이터 전송은 HTTP Body 를 통해 전달하는 것으로 알고 있습니다. 혹시 쿼리 파라미터로 전달하는 이유가 있을까요?

dkdud9261 replied:
chatGPT에서 쿼리 파라미터를 사용하는 예시를 보고 간단한 Long 데이터 하나라 사용해봤는데, 보편성과 통일성을 고려해 update를 위한 데이터 전송 시에는 RequestBody를 사용하는 것으로 통일했습니다 !

PPakSang commented on ilkda/src/main/java/com/ilkda/server/record/dto/RecordTextForm.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+public class RecordTextForm {
+
+    private String text;
+}
```
RecordTextForm 을 @RequestBody 로 요청을 받았을 때 정상적으로 테스트가 수행되나요? 기본 생성자를 추가하거나, 단일 파라미터 생성자는 @JsonCreator 로 명시하지 않으면 오류가 나는 것으로 알고 있습니다.

dkdud9261 replied:
기본 생성자 추가로 오류 수정했습니다~!

PPakSang commented on ilkda/src/main/java/com/ilkda/server/record/dto/RegisterRecordForm.java:
```diff
@@ -2,9 +2,11 @@
 import lombok.Data;
+import lombok.NoArgsConstructor;
 
 @Data
 @AllArgsConstructor
+@NoArgsConstructor
 public class RegisterRecordForm {
 
     private Long bookId;
```
위에서 언급한 문제를 해결하신 부분으로 보입니다. 사용하지 않는 생성자라면 접근 레벨을 private 으로 설정하는 것이 나을 것 같습니다!

dkdud9261 replied:
수정했습니다~!

PPakSang commented:
1. RecordService Class 내에 Validation 로직이 많이 추가됐다는 느낌이 드네요. 따로 RecordValidator 클래스로 메소드를 추출하는 것도 나쁘지 않을 것 같습니다. 

2. 현재 Record 데이터를 사용하는 과정에서 거의 필수적으로 Book 을 참조하게 되는데, fetch join 으로 쿼리를 수정해 두 개 이상의 쿼리가 전송되지 않도록 하는 것도 고려해보면 좋을 것 같습니다.

dkdud9261 committed:
Refactor: 피드백 반영

- 인증된 사용자 정보 조회 로직 수정
- update를 위한 데이터 전송 requst body로 통일
- 입력 데이터 클래스에 기본 생성자 추가, private으로 설정
- service 내 validation 로직 model 클래스로 이동
- Record에서 Book 참조 시 fetch join 사용

PPakSang commented:
피드백 반영한 파일을 개별 커밋으로 나누고 Conversation 에 커밋 참조 걸어주시면 수정 내용 확인하기가 더 쉬울 것 같습니다.

고생하셨습니다👍
